gamemodes/terrortown/gamemode/server/sv_roleselection.lua

```diff
@@ -0,0 +1,496 @@
+roleselection.cv.ttt_max_baseroles_pct = CreateConVar("ttt_max_baseroles_pct", "0", {FCVAR_NOTIFY, FCVAR_ARCHIVE}, "Maximum amount of different baseroles based on player amount. ttt_max_baseroles needs to be 0")
+
+---
+-- Returns the amount of pre selected @{ROLE}s
+-- @param number subrole subrole id of a @{ROLE}'s index
+-- @return number amount
+-- @realm server
+function roleselection.GetPreSelectedRole(subrole)
+	local tmp = 0
+
+	if GetRoundState() == ROUND_ACTIVE then
+		local plys = player.GetAll()
+
+		for i = 1, #plys do
+			local ply = plys[i]
+
+			if not IsValid(ply) or ply:GetForceSpec() or ply:GetSubRole() ~= subrole then continue end
+
+			tmp = tmp + 1
+		end
+	elseif roleselection.finalRoles then
+		for ply, sr in pairs(roleselection.finalRoles) do
+			if sr == subrole then
+				tmp = tmp + 1
+			end
+		end
+	end
+
+	return tmp
+end
+
+local function GetAvailableRoleAmount(roleData, forced, maxPlys)
+	local bool = true
+
+	if not forced then
+		local randomCVar = GetConVar("ttt_" .. roleData.name .. "_random")
+
+		bool = math.random(100) <= (randomCVar and randomCVar:GetInt() or 100)
+	end
+
+	if bool then
+		return roleData:GetRoleCount(maxPlys) - roleselection.GetPreSelectedRole(roleData.index)
+	end
+
+	return 0
+end
+
+---
+-- Returns a list of all selectable players
+-- @param table plys list of @{Player}s that should be filtered.
+-- @return table list of filtered players
+-- @realm server
+function roleselection.GetSelectablePlayers(plys)
+	local tmp = {}
+
+	for i = 1, #plys do
+		local ply = plys[i]
+
+		-- everyone on the spec team is in specmode
+		if not ply:GetForceSpec() and not hook.Run("TTT2DisableRoleSelection", ply) then
+			tmp[#tmp + 1] = ply
+		end
+	end
+
+	return tmp
+end
+
+---
+-- Returns all selectable @{ROLE}s based on the amount of @{Player}s
+-- @param number maxPlys amount of maximum @{Player}s (filtered!)
+-- @return table a list of all selectable @{ROLE}s
+-- @realm server
+function roleselection.GetAllSelectableRolesList(maxPlys)
+	if maxPlys < 2 then return end
+
+	local forcedRolesTbl = {}
+
+	for _, subrole in pairs(roleselection.forcedRoles) do
+		forcedRolesTbl[subrole] = true
+	end
+
+	local rolesCountTbl = {
+		[INNOCENT] = GetAvailableRoleAmount(INNOCENT, true, maxPlys),
+		[TRAITOR] = GetAvailableRoleAmount(TRAITOR, true, maxPlys)
+	}
+
+	local checked = {}
+	local rlsList = roles.GetList()
+
+	-- add all roles if possible
+	for i = 1, #rlsList do
+		local roleData = rlsList[i]
+
+		if checked[roleData.index] then continue end
+
+		checked[roleData.index] = true
+
+		-- INNOCENT and TRAITOR are all the time selectable
+		if roleData == INNOCENT or roleData == TRAITOR or not roleData:IsSelectable() then continue end
+
+		-- if this is a subrole (a role that have a baserole is a subrole), check if the base role is available first
+		if roleData.baserole and roleData.baserole ~= ROLE_INNOCENT and roleData.baserole ~= ROLE_TRAITOR then
```

Comment:
maybe this is a dumb idea, but wouldn't it be better to ignore non base roles here completely and select possible subroles only in a later step?

Reply:
This step there is needed to match all available roles. This has nothing to do with the limitation or filtering. If there is a subrole, we need the baserole at first. If this baserole isn't available, we don't need to select any subrole. In that way we save loops.